### PR TITLE
Changes to implement SingleStore JWKS logic

### PR DESCRIPTION
**Pull request by pmishchenko-ua**
This PR is created for visibility of changes made to the original repo.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,10 +1,10 @@
-[![Go Report Card](https://goreportcard.com/badge/github.com/MicahParks/keyfunc)](https://goreportcard.com/report/github.com/MicahParks/keyfunc) [![Go Reference](https://pkg.go.dev/badge/github.com/MicahParks/keyfunc.svg)](https://pkg.go.dev/github.com/MicahParks/keyfunc)
+[![Go Report Card](https://goreportcard.com/badge/github.com/pmishchenko-ua/jwt-keyfunc)](https://goreportcard.com/report/github.com/pmishchenko-ua/jwt-keyfunc) [![Go Reference](https://pkg.go.dev/badge/github.com/pmishchenko-ua/jwt-keyfunc.svg)](https://pkg.go.dev/github.com/pmishchenko-ua/jwt-keyfunc)
```
Comment (muir) [inline README.md:1]: These are now `memsql/jwt-keyfunc` right?

File: checksum_test.go [hunk 2]
```diff
@@ -2,21 +2,22 @@ package keyfunc_test
...
+	"github.com/golang-jwt/jwt"
 
-	"github.com/MicahParks/keyfunc"
+	keyfunc "github.com/pmishchenko-ua/jwt-keyfunc"
 )
 
 // TestChecksum confirms that the JWKS will only perform a refresh if a new JWKS is read from the remote resource.
 func TestChecksum(t *testing.T) {
-	tempDir, err := os.MkdirTemp("", "*")
+	tempDir, err := ioutil.TempDir("", "*")
```
Comment (muir) [inline checksum_test.go:20]: Why did you change this?

File: checksum_test.go [hunk 3]
```diff
@@ -29,7 +30,7 @@ func TestChecksum(t *testing.T) {
 
 	jwksFile := filepath.Join(tempDir, jwksFilePath)
 
-	err = os.WriteFile(jwksFile, []byte(jwksJSON), 0600)
+	err = ioutil.WriteFile(jwksFile, []byte(jwksJSON), 0600)
```
Comment (muir) [inline checksum_test.go:33]: Ditto

File: get.go [hunk 4]
```diff
@@ -181,15 +181,18 @@ func (j *JWKS) refresh() (err error) {
...
+				if kidIdx >= 0 {
 					continue
 				}
 			}
-
-			j.keys[kid] = parsedJWK{public: key.inter}
+			if kidIdx >= 0 {
+				j.keys[kidIdx] = ParsedJWK{Public: key.inter, kid: kid}
+			} else {
+				j.keys = append(j.keys, ParsedJWK{Public: key.inter, kid: kid})
```
Comment (muir) [inline get.go]:
```suggestion
				j.keys[kidIdx] = append(j.keys[kidIdx], ParsedJWK{Public: key.inter, kid: kid})
```
Easy to do `map[string][]key` ...

File: given.go [hunk 5]
```diff
@@ -6,6 +6,12 @@ import (
 	"crypto/rsa"
 )
 
+// GivenKey represents a cryptographic key that resides in a JWKS. In conjuncture with Options.
```
Comment (muir) [inline given.go]: Extras, please remove

File: jwks.go [hunk 6]
```diff
@@ -180,22 +191,84 @@ func (j *JWKS) RawJWKS() []byte {
...
+	// jwkUseWhitelist might be empty if the jwks was from keyfunc.NewJSON() or if JWKUseNoWhitelist option was true.
+	// in this case we don't restrict "use" parameter
+	if len(j.jwkUseWhitelist) > 0 {
+		_, canUseKey = j.jwkUseWhitelist[key.use]
+	}
+	return canUseKey
+}
+
+func (j *JWKS) getMatchingKeys(alg, kid, iss string) []*ParsedJWK {
+	result := make([]*ParsedJWK, 0, 1)
```
Comment (muir) [inline jwks.go]:
```suggestion
	var result []*ParsedJWK
```
What you've got is fine, but I don't think it has any advantage over the simpler declaration above

File: jwks.go [hunk 7]
```diff
@@ -180,22 +191,84 @@ func (j *JWKS) RawJWKS() []byte {
...
+	if kid != "" {
+		for idx, key := range j.keys {
+			if (key.kid == kid) && (key.algorithm == alg || key.algorithm == "") && j.canUseKey(key) {
+					result = append(result, &j.keys[idx])
```
Comment (muir) [inline jwks.go]: indentation isn't right.  Have you run `gofmt`?

Comment (muir) [inline jwks.go]:
```suggestion
					result = append(result, key)
```
I thought keys were already pointers?
If not, the simpler thing is to also say: `key := key` and that breaks the loop aliasing.

File: jwks.go [hunk 8]
```diff
@@ -180,22 +191,84 @@ func (j *JWKS) RawJWKS() []byte {
...
+//
+// 1. If the JWT has a kid (Key ID) field, the JWKs with matching kid fields are validated.
+// 2. If the JWT has a kid field that doesn’t match any JWK or jwt_config key, the authentication request is rejected. See Validate JWTs with the jwt-config for more information.
+// 3. If the JWT has an iss (Issuer) field (instead of a kid field) that matches the kid in one or more JWKs, the JWKs with matching kid fields are validated.
+// 4. If the JWT does not have a kid field and the iss field does not match the kid field in any JWK, then validation is attempted with all the JWKs with a matching alg (Algorithm) field. If the alg field is not specified, the kty (Key Type) field is used instead.
+//
+func (j *JWKS) GetMatchingKeysWithRefresh(alg, kid, iss string) (matchingKeys []*ParsedJWK, err error) {
 	j.mux.RLock()
-	pubKey, ok := j.keys[kid]
+	matchingKeys = j.getMatchingKeys(alg, kid, iss)
```
Comment (muir) [inline jwks.go]:
```suggestion
        func() {
             j.mux.RLock()
             defer j.mux.RUnlock()
	     matchingKeys = j.getMatchingKeys(alg, kid, iss)
	}()
```

Unless it's really hard, unlock should be done with `defer` so that an un-expected panic doesn't lead to program hangs

File: README.md (hunk 1 above)
Comment (pmishchenko-ua) [reply README.md:1]: right, I will rename references in the next commit in this repo.

File: checksum_test.go (hunk 3 above)
Comment (pmishchenko-ua) [reply checksum_test.go:33]: go 1.15 compatibility (it is used in engine build) - `os.WriteFile` doesn't exist there

File: checksum_test.go (hunk 2 above)
Comment (pmishchenko-ua) [reply checksum_test.go:20]: go 1.15 compatibility

File: jwks.go (hunk 7 above)
Comment (pmishchenko-ua) [reply jwks.go]: `key := key` looks nice, the reason for `&j.keys[idx]` was indeed loop aliasing.

Comment (pmishchenko-ua) [reply jwks.go]: I did, but not with the latest update, thanks for pointing out.

File: jwks.go (hunk 8 above)
Comment (pmishchenko-ua) [reply jwks.go]: makes sense, I'm moving lock and unlock to `getMatchingKeys` itself as it's always invoked with it.

File: checksum_test.go (hunk 2 above)
Comment (muir) [reply checksum_test.go:20]: There are other versions of go available in the engine repo, I'm pretty sure.

File: get.go (hunk 4 above)
Comment (pmishchenko-ua) [reply get.go]: makes sense, changed to map

File: README.md [hunk 9]
```diff
@@ -48,21 +48,21 @@ this Go package, please open an issue or pull request.
 For complete examples, please see the `examples` directory.
 
 ```go
-import "github.com/MicahParks/keyfunc"
+import keyfunc "github.com/pmishchenko-ua/jwt-keyfunc"
 ```
 
 #### A note on read-only keys
-The [`JWKS.ReadOnlyKeys`](https://pkg.go.dev/github.com/MicahParks/keyfunc#JWKS.ReadOnlyKeys) method returns a read-only
+The [`JWKS.ReadOnlyKeys`](https://pkg.go.dev/github.com/pmishchenko-ua/jwt-keyfunc#JWKS.ReadOnlyKeys) method returns a read-only
```
Comment (carlsverre) [inline README.md:55]: update all these links to point at singlestore-labs rather than your account

File: alg_test.go [hunk 10]
```diff
@@ -5,9 +5,9 @@ import (
 	"errors"
 	"testing"
 
-	"github.com/golang-jwt/jwt/v4"
+	"github.com/golang-jwt/jwt"
 
-	"github.com/MicahParks/keyfunc"
+	keyfunc "github.com/pmishchenko-ua/jwt-keyfunc"
```
Comment (carlsverre) [inline alg_test.go:10]: oh man... this is very risky - how to guarantee you update all these imports before landing?

File: get.go [hunk 11]
```diff
@@ -59,6 +60,12 @@ func Get(jwksURL string, options Options) (jwks *JWKS, err error) {
 // backgroundRefresh is meant to be a separate goroutine that will update the keys in a JWKS over a given interval of
 // time.
 func (j *JWKS) backgroundRefresh() {
+	if j.initAsync {
+		if err := j.refresh(); err != nil && j.refreshErrorHandler != nil {
+			j.refreshErrorHandler(err)
+		}
```
Comment (carlsverre) [inline get.go]: This makes no sense to me - the point of this function seems to be running `j.refresh`... why are you running it right here? Was this for testing?

Comment (carlsverre) [reply get.go]: Is the issue that if initAsync is set then we want the background refresh to kick off right away rather than waiting for the initial interval?

In this case, it seems you could just write to `j.refreshRequests` after starting the `backgroundRefresh` goroutine in the `Get` function.

File: README.md (hunk 9 above)
Comment (pmishchenko-ua) [reply README.md:55]: will update the links (they are in tests and examples only) in a subsequent commit in the singlestore-labs repo. The package itself will be valid anyway.

File: alg_test.go (hunk 10 above)
Comment (pmishchenko-ua) [reply alg_test.go:10]: I will update them after landing - links in this repo are in README, examples and tests, so not a big risk.

File: get.go (hunk 11 above)
Comment (pmishchenko-ua) [reply get.go]: good idea, I added writing to `backgroundRefresh`

**Review by carlsverre — APPROVED** [review]
NICE

File: jwks.go [hunk 12]
```diff
@@ -177,60 +204,194 @@ func (j *JWKS) RawJWKS() []byte {
...
+// GetMatchingKeys implements the logic described in
+// https://docs.singlestore.com/db/v7.8/en/security/authentication/authenticate-via-jwt.html
+// A Read Lock for `j.mux` is acquired when the JWKS is read
+//
+// JWTs are matched with JSON Web Keys (JWKs) for validation as follows:
+// 1. If the JWT has a kid (Key ID) field, the JWKs with matching kid fields are validated.
+// 2. If the JWT has a kid field that doesn’t match any JWK or jwt_config key, the authentication request is rejected. See Validate JWTs with the jwt-config for more information.
+// 3. If the JWT has an iss (Issuer) field (instead of a kid field) that matches the kid in one or more JWKs, the JWKs with matching kid fields are validated.
+// 4. If the JWT does not have a kid field and the iss field does not match the kid field in any JWK, then validation is attempted with all the JWKs with a matching alg (Algorithm) field. If the alg field is not specified, the kty (Key Type) field is used instead.
+func (j *JWKS) GetMatchingKeys(token *jwt.Token) ([]*ParsedJWK, error) {
```
Comment (gzharun-ua) [inline jwks.go:309]: @pmishchenko-ua, this function looks LGTM.
Please, proceed.
